File: bindings/python/tests/test_runtime.py
"""Runtime surface: the process-wide Go knobs (heap limit, GC
percentage, GOMAXPROCS, heap profile, pool counters) and the shipped
inner-hash registry enumeration."""

from __future__ import annotations

import os
import tempfile
import unittest

import itb3 as itb


class RuntimeTest(unittest.TestCase):
    def test_set_gomaxprocs_queries_then_restores(self) -> None:
        # A non-positive argument queries without changing; the setter
        # returns the value that was in force before it.
        before = itb.set_gomaxprocs(0)
        self.assertGreater(before, 0)
        self.assertEqual(itb.set_gomaxprocs(2), before)
        self.assertEqual(itb.set_gomaxprocs(0), 2)
        itb.set_gomaxprocs(before)
        self.assertEqual(itb.set_gomaxprocs(0), before)

    def test_write_heap_profile_writes_a_readable_profile(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            path = os.path.join(tmp, "heap.prof")
            itb.write_heap_profile(path)
            self.assertGreater(os.path.getsize(path), 0)
            with open(path, "rb") as f:
                # pprof profiles are gzip-wrapped protobuf.
                self.assertEqual(f.read(2), b"\x1f\x8b")

    def test_write_heap_profile_reports_the_os_diagnostic(self) -> None:
        with self.assertRaises(itb.ItbError) as ctx:
            itb.write_heap_profile("/no-such-directory-itb3-test/heap.prof")
        self.assertEqual(ctx.exception.status, itb.Status.BAD_INPUT)
        self.assertIn("heap.prof", ctx.exception.message)

    def test_pool_stats_length_matches_the_declared_layout(self) -> None:
        length = itb.pool_stats_len()
        self.assertGreater(length, 0)
        stats = itb.pool_stats()
        self.assertEqual(len(stats), length)
        tiers = stats[0]
        # Slot 0 carries the tier count T; the vector is 1 + 5*T + 8.
        self.assertGreater(tiers, 0)
        self.assertEqual(length, 1 + 5 * tiers + 8)

    def test_pool_stats_counters_are_monotonic_across_work(self) -> None:
        before = itb.pool_stats()
        with itb.Pipeline.init("singlemsg-triple-mac-v1") as pipe:
            pipe.decrypt_message(pipe.encrypt_message(b"x" * 4096))
        after = itb.pool_stats()
        self.assertEqual(len(after), len(before))
        for i in range(1, len(after)):
            self.assertGreaterEqual(after[i], before[i])
        self.assertGreater(sum(after[1:]), sum(before[1:]))

    def test_memory_limit_and_gc_percent_query_without_changing(self) -> None:
        limit = itb.set_memory_limit(-1)
        self.assertEqual(itb.set_memory_limit(-1), limit)
        pct = itb.set_gc_percent(-1)
        self.assertEqual(itb.set_gc_percent(-1), pct)


class HashRegistryTest(unittest.TestCase):
    def test_hash_names_enumerates_the_shipped_registry(self) -> None:
        names = itb.hash_names()
        self.assertGreater(len(names), 1)
        self.assertEqual(len(names), len(set(names)))
        for name in names:
            self.assertIsInstance(name, str)
            self.assertTrue(name)
        # The enumeration is what a caller validates a primitive name
        # against, so a shipped name resolves and a typo does not.
        self.assertIn("areion512", names)
        self.assertNotIn("areion512-nope", names)

    def test_every_enumerated_name_constructs_a_pipeline(self) -> None:
        for name in itb.hash_names():
            opts = itb.Opts().with_inner_hash(name).with_parallax(False)
            with itb.Pipeline.init("singlemsg-triple-nomac-v1", opts) as pipe:
                self.assertEqual(
                    pipe.decrypt_message(pipe.encrypt_message(b"registry probe")),
                    b"registry probe",
                )


if __name__ == "__main__":
    unittest.main()

File: bindings/cpp/tests/test_runtime.cpp
/* The runtime surface the stress harness drives: GOMAXPROCS, the heap
 * profile, the pool counters, and the shipped hash-registry
 * enumeration. */

#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>

#include "test_util.hpp"

static int run()
{
    /* GOMAXPROCS: zero queries, a positive value sets and returns the
     * previous one. The original value is restored so a later test in
     * the same binary is unaffected. */
    const int original = itb::set_gomaxprocs(0);
    TEST_ASSERT(original > 0, "query must report a positive GOMAXPROCS, got %d", original);
    const int previous = itb::set_gomaxprocs(2);
    TEST_ASSERT(previous == original, "set must return the previous value, %d != %d",
                previous, original);
    TEST_ASSERT(itb::set_gomaxprocs(0) == 2, "GOMAXPROCS must read back as 2");
    (void)itb::set_gomaxprocs(original);

    /* Pool counters: the slot count comes from the library, the layout
     * is 1 + 5*T + 8 with T in slot 0, and every counter is a
     * non-negative monotonic total. */
    const std::size_t slots = itb::pool_stats_len();
    TEST_ASSERT(slots >= 9, "pool_stats_len must report at least 9 slots, got %zu", slots);
    std::vector<std::int64_t> counters(slots);
    const std::size_t written = itb::pool_stats(counters);
    TEST_ASSERT(written == slots, "pool_stats wrote %zu of %zu slots", written, slots);
    const std::int64_t tiers = counters[0];
    TEST_ASSERT(tiers > 0, "slot 0 must carry the tier count, got %lld",
                static_cast<long long>(tiers));
    TEST_ASSERT(static_cast<std::size_t>(1 + 5 * tiers + 8) == slots,
                "layout 1 + 5*%lld + 8 != %zu", static_cast<long long>(tiers), slots);
    for (std::size_t i = 0; i < slots; i++) {
        TEST_ASSERT(counters[i] >= 0, "counter %zu is negative", i);
    }

    /* A destination shorter than the requirement is a relayed
     * BufferTooSmall, not a silent truncation. */
    bool too_small = false;
    try {
        std::vector<std::int64_t> narrow(1);
        (void)itb::pool_stats(narrow);
    } catch (const itb::Error &e) {
        too_small = e.status() == itb::Status::BufferTooSmall;
    }
    TEST_ASSERT(too_small, "a short destination must raise BufferTooSmall");

    /* Heap profile: a writable path yields a non-empty pprof file
     * whose gzip magic is intact; an unwritable one throws. */
    const std::string path = std::string("/tmp/itb-cpp-runtime-test-") +
                             std::to_string(static_cast<long long>(itb::pool_stats_len())) +
                             ".pprof";
    itb::write_heap_profile(path);
    std::FILE *f = std::fopen(path.c_str(), "rb");
    TEST_ASSERT(f != nullptr, "heap profile was not created at %s", path.c_str());
    unsigned char magic[2] = {0, 0};
    const std::size_t got = std::fread(magic, 1, sizeof(magic), f);
    std::fclose(f);
    std::remove(path.c_str());
    TEST_ASSERT(got == 2 && magic[0] == 0x1f && magic[1] == 0x8b,
                "heap profile is not a gzip stream");

    bool rejected = false;
    try {
        itb::write_heap_profile("/proc/itb-no-such-directory/heap.pprof");
    } catch (const itb::Error &) {
        rejected = true;
    }
    TEST_ASSERT(rejected, "an unwritable path must throw");

    /* Hash registry: a JSON array of strings carrying the names
     * Pipeline::init accepts, with the canonical first entry present. */
    const std::string names = itb::hash_names();
    TEST_ASSERT(names.size() > 2 && names.front() == '[' && names.back() == ']',
                "hash_names must be a JSON array, got %s", names.c_str());
    TEST_ASSERT(names.find("\"areion512\"") != std::string::npos,
                "hash_names must carry areion512");
    TEST_ASSERT(names.find("\"aesitb128\"") != std::string::npos,
                "hash_names must carry aesitb128");
    TEST_ASSERT(names.find("\"nope\"") == std::string::npos,
                "hash_names must not carry an unregistered name");

    /* The enumeration is the authority the flag validation reads: a
     * name it carries constructs, one it does not is rejected. */
    itb::Opts opts;
    opts.set("innerHash", "blake3");
    itb::Pipeline p = itb::Pipeline::init("singlemsg-triple-mac-v1", opts);
    TEST_ASSERT(!p.save().empty(), "a registry name must construct");
    return 0;
}

TEST_MAIN(run)
